server/src/main/java/org/opensearch/index/translog/RemoteFsTranslogWithPinnedTimestamps.java

```diff
@@ -0,0 +1,378 @@
+
+                    // If index is not deleted, make sure to keep latest metadata file
+                    if (indexDeleted == false) {
+                        metadataFilesToBeDeleted.remove(metadataFiles.get(0));
+                    }
+
+                    if (metadataFilesToBeDeleted.isEmpty()) {
+                        logger.debug("No metadata files to delete");
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                        return;
+                    }
+
+                    logger.debug("metadataFilesToBeDeleted = {}", metadataFilesToBeDeleted);
+                    // For all the files that we are keeping, fetch min and max generations
+                    List<String> metadataFilesNotToBeDeleted = new ArrayList<>(metadataFiles);
+                    metadataFilesNotToBeDeleted.removeAll(metadataFilesToBeDeleted);
+
+                    logger.debug("metadataFilesNotToBeDeleted = {}", metadataFilesNotToBeDeleted);
+                    Set<Long> generationsToBeDeleted = getGenerationsToBeDeleted(
+                        metadataFilesNotToBeDeleted,
+                        metadataFilesToBeDeleted,
+                        indexDeleted
+                    );
+
+                    logger.debug("generationsToBeDeleted = {}", generationsToBeDeleted);
+                    if (generationsToBeDeleted.isEmpty() == false) {
+                        // Delete stale generations
+                        translogTransferManager.deleteGenerationAsync(
+                            primaryTermSupplier.getAsLong(),
+                            generationsToBeDeleted,
+                            remoteGenerationDeletionPermits::release
+                        );
+
+                        // Delete stale metadata files
+                        translogTransferManager.deleteMetadataFilesAsync(
+                            metadataFilesToBeDeleted,
+                            remoteGenerationDeletionPermits::release
+                        );
+
+                        // Update cache to keep only those metadata files that are not getting deleted
+                        oldFormatMetadataFileGenerationMap.keySet().retainAll(metadataFilesNotToBeDeleted);
+
+                        // Delete stale primary terms
+                        deleteStaleRemotePrimaryTerms(metadataFiles);
+                    } else {
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                    }
+                } catch (Exception e) {
+                    remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                logger.error("Exception while listing translog metadata files", e);
+            }
```

Comment:
good to add assertions on all permits

Reply:
The problem is these permits will only be released post generations and metadata delete. So, we won't be able to check if the permits in the end are available or not.